doc/development_guide/profiling.rst

```diff
@@ -0,0 +1,104 @@
+.. -*- coding: utf-8 -*-
+.. _profiling:
+
+===================================
+ Profiling and performance analysis
+===================================
+
+Performance analysis for Pylint
+-------------------------------
+
+To analyse the performance of Pylint we recommend to use the ``cProfile`` module
+from ``stdlib``. Together with the ``pstats`` module this should give you all the tools
+you need to profile a Pylint run and see which functions take how long to run.
+
+The documentation for both modules can be found at cProfile_.
+
+To profile a run of Pylint over itself you can use the following code and run it from the base directory.
+Note that ``cProfile`` will create a document called ``stats`` that is then read by ``pstats``. The
+human-readable output will be stored by ``pstats`` in ``./profiler_stats``. It will be sorted by
+``cumulative time``:
+
+.. sourcecode:: python
+
+    import cProfile
+    import pstats
+    import sys
+
+    sys.argv = ["pylint", "pylint"]
+    cProfile.run("from pylint import __main__", "stats")
+
+    with open("profiler_stats", "w", encoding="utf-8") as file:
+        stats = pstats.Stats("stats", stream=file)
+        stats.sort_stats("cumtime")
+        stats.print_stats()
+
+You can also interact with the stats object by sorting or restricting the output.
+For example, to only print functions from the ``pylint`` module and sort by cumulative time you could
+use:
+
+.. sourcecode:: python
+
+    import cProfile
+    import pstats
+    import sys
+
+    sys.argv = ["pylint", "pylint"]
+    cProfile.run("from pylint import __main__", "stats")
+
+    with open("profiler_stats", "w", encoding="utf-8") as file:
+        stats = pstats.Stats("stats", stream=file)
+        stats.sort_stats("cumtime")
+        stats.print_stats("pylint/pylint")
+
+Lastly, to profile a run over your own module or code you can use:
+
+.. sourcecode:: python
+
+    import cProfile
+    import pstats
+    import sys
+
+    sys.argv = ["pylint", "your_dir/your_file"]
+    cProfile.run("from pylint import __main__", "stats")
+
+    with open("profiler_stats", "w", encoding="utf-8") as file:
+        stats = pstats.Stats("stats", stream=file)
+        stats.sort_stats("cumtime")
+        stats.print_stats()
+
+The documentation of the ``pstats`` module discusses other possibilites to interact with
+the profiling output.
+
+
+Performance analysis of a specific checker
+------------------------------------------
+
+To analyse the performance of specific checker within Pylint we can use the human-readable output
+created by ``pstats``.
+
+If you search in the ``profiler_stats`` file for the file name of the checker you will find all functional
```

Comment:
Maybe we could also talk about some tools that interpret it like ``gprof2dot`` or ``snakeviz`` (you don't have to analyses the text output yourself even if you can) ?

Reply:
That's a good idea, but I would keep some explanation (or link to existing documentation) on how to interact, sort and filter the ``pstats`` file as well. From the discussion in #4814 we saw that the graph for ``pylint`` can become very large and hard to follow. If I were to analyse performance issues in a checker, I would probably work more with the ``Stats`` object than the graph.

Reply:
Yes pre-sorting with pstat is really useful.

Reply:
@Pierre-Sassoulas The original issue was created by an user wondering why their program took so long too lint. That was what prompted me to include basic advice as well. Do you want me to remove this?

I'll create a section about interacting with `gprof2dot` and pre-sorting. I don't want this to be too long: if people want to learn how to profile they should look elsewhere, but this documentation should set them on the right track with regards to any `pylint` specifics.

Reply:
@DudeNr33 Did you figure out how to use `print_stats("pylint/pylint")` with `gprof2dot`? I don't think `print_stats` changes the `pstats` `stats` object itself, so I can't restrict the file passed to `gprof2dot`

Reply:
I don't know if this would make sense, as the diagram is a call graph and filtering would lead to incomplete or maybe even broken graphs. 
You could probably directly delete entries from the ``stats`` attribute and then use [``dump_stats()``](https://stackoverflow.com/questions/51536411/saving-cprofile-results-to-readable-external-file) to save it. I don't know however how the resulting output with ``gprof2dot`` looks, and if it is even able to process it properly if some nodes in the graph are missing.

Reply:
Hmm okay, so a short section pointing to `gprof2dot` would be enough?

Reply:
I think so. As Pierre pointed out ``snakeviz`` is an alternative visualisation tool ([this article](https://medium.com/@narenandu/profiling-and-visualization-tools-in-python-89a46f578989) provides a brief comparison of the tools) which provides a different representation together with filtering and sorting of the output, so I think this should be included, too.

Reply:
> @Pierre-Sassoulas The original issue was created by an user wondering why their program took so long too lint. That was what prompted me to include basic advice as well. Do you want me to remove this?

> if people want to learn how to profile they should look elsewhere

Il agree let's keep the basic, but also let's keep it basic :) a link to popular tools' docs are enough imo.